code/datums/components/harvester.dm

```diff
@@ -0,0 +1,142 @@
+
+/datum/component/harvester/proc/attackby(datum/source, obj/item/I, mob/user)
+	if(user.do_actions)
+		return TRUE
+
+	if(!isreagentcontainer(I) || istype(I, /obj/item/reagent_containers/pill))
+		to_chat(user, span_rose("[I] isn't compatible with [source]."))
+		return TRUE
+
+	var/trans
+	var/obj/item/reagent_containers/container = I
+
+	if(!container.reagents.total_volume)
+		trans = beaker.reagents.trans_to(container, 30)
+		to_chat(user, span_rose("[trans ? "You take [trans]u out of the internal storage. It now contains [beaker.reagents.total_volume]u" : "[source]'s storage is empty."]."))
+		return TRUE
+
+	if(length(container.reagents.reagent_list) > 1)
+		to_chat(user, span_rose("The solution needs to be uniform and contain only a single type of reagent to be compatible."))
+		return TRUE
+
+	if(beaker.reagents.total_volume && (container.reagents.reagent_list[1].type != beaker.reagents.reagent_list[1].type))
+		to_chat(user, span_rose("[source]'s internal storage can contain only one kind of solution at the same time. It currently contains <b>[beaker.reagents.reagent_list[1].name]</b>"))
+		return TRUE
+
+	if(!locate(container.reagents.reagent_list[1].type) in loadable_reagents)
+		to_chat(user, span_rose("This reagent is not compatible with the weapon's mechanism. Check the engraved symbols for further information."))
+		return TRUE
+
+	if(container.reagents.total_volume < 5)
+		to_chat(user, span_rose("At least 5u of the substance is needed."))
+		return TRUE
+
+	if(beaker.reagents.total_volume >= 30)
+		to_chat(user, span_rose("The internal storage is full."))
+		return TRUE
+
+	to_chat(user, span_notice("You begin filling up the [source] with [container.reagents.reagent_list[1]]."))
+	if(!do_after(user, 1 SECONDS, TRUE, source, BUSY_ICON_BAR, null, PROGRESS_BRASS))
+		return TRUE
+
+	trans = container.reagents.trans_to(beaker, container.amount_per_transfer_from_this)
+	to_chat(user, span_rose("You load [trans]u into the internal system. It now holds [beaker.reagents.total_volume]u."))
+	return TRUE
+
+/datum/component/harvester/proc/unique_action(datum/source, mob/user)
+	if(loaded_reagent)
+		to_chat(user, span_rose("The blade is powered with [loaded_reagent.name]. You can release the effect by stabbing a creature."))
+		return FALSE
+
+	if(beaker.reagents.total_volume < 5)
+		to_chat(user, span_rose("You don't have enough substance."))
+		return FALSE
+
+	if(user.do_actions)
+		return FALSE
+
+	to_chat(user, span_rose("You start filling up the small chambers along the blade's edge."))
+	if(!do_after(user, 2 SECONDS, TRUE, source, BUSY_ICON_BAR, ignore_turf_checks = TRUE))
+		to_chat(user, span_rose("Due to the sudden movement, the safety machanism drains out the reagent back into the main storage."))
+		return FALSE
+
+	loaded_reagent = beaker.reagents.reagent_list[1]
+	beaker.reagents.remove_any(5)
+	return TRUE
+
+/datum/component/harvester/proc/apply_flame_cone(mob/living/target, mob/living/user)
```

Comment:
this is a useless proc. In byond we don't split procs in small bit unless it's really necessary, because the proc overhead cost is huge.

If you don't want to repeat code, make it a macro